### ISSUE TITLE
Consider OffsetTime and OffsetDate types

### ISSUE THREAD

**Issue by mj1856** [nodatime/nodatime#864]
Not saying per se that I want this, but there may be a need for an `OffsetTime` type - that is, a type that is `LocalTime` + `Offset`.

https://stackoverflow.com/questions/44514504/offsettime-in-nodatime

ISO8601 has this in section *4.2.5.2 Local time and the difference from UTC*, supporting the same basic/extended formats that we are used to with `OffsetDateTime`, sans the date portion.

Java 8 has this with [`java.time.OffsetTime`](https://docs.oracle.com/javase/8/docs/api/java/time/OffsetTime.html).

PostgreSQL has this with [`time with time zone`](https://www.postgresql.org/docs/current/static/datatype-datetime.html), and mentions that it's in the SQL standard (ISO 9075), but discourages its use.  From their docs:

> The type `time with time zone` is defined by the SQL standard, but the definition exhibits properties which lead to questionable usefulness. In most cases, a combination of date, time, timestamp without time zone, and timestamp with time zone should provide a complete range of date/time functionality required by any application.

Do we want to add this to Noda Time?

**Comment by mj1856** [comment]
We should also consider `OffsetDate`.  Though it's an odd duck, we have seen at least [one StackOverflow question on this](https://stackoverflow.com/q/44409687/634824).

In XML Schema Definition, both `xs:date` and `xs:time` have an offset field.  Though the offset is optional, it is reasonable that one might want to round-trip such a value through an object of some sort.

Alternatively, we could make an offset an option on the `LocalDate` and `LocalTime` objects, but I think that goes against our stated design philosophy.

**Comment by mj1856** [comment]
Note that for `xs:date`, the offset is intended to reflect that on the first instant of the date.  If the offset changes over the date, that is not reflected.

Similarly, `xs:time` doesn't take into account that the offset could be different if applied to a different date.  Its use is limited, but is something to the effect of a recurring appointment in a fixed-offset time zone.

**Comment by jskeet** [comment]
Okay - will start looking at this. We should probably think about what we want to do in terms of serialization (XML, binary, JSON) - that can come once we've got the basic functionality ready.

**Comment by mj1856** [comment]
XML should be as specified for `xs:date` and `xs:time`.  The `xs:time` spec is ISO-8601 compliant, as in `23:59:59Z` or `23:59:59+05:30`.   The `xs:date` is similar (though *not* ISO-8601 compliant), as in `2017-12-31Z` or `2017-12-31+05:30`.

I think it's reasonable for JSON to match.  Not sure about binary.

**Comment by mj1856** [comment]
Just ran across another use case fore `DateOffset`.  It's reasonable to take timestamps of events, logs, etc. by UTC.  When one has a lot of these, often they are grouped by date.  In that situation, it seems reasonable to label the groups/folders/buckets to be clear that they are the UTC date, as in `2017-11-01Z`, `2017-11-02Z`, etc.

**Comment by jskeet** [comment]
This is presumably still open because there's something I still need to do. I suspect it's documentation for the serialization...

**Comment by mqudsi** [comment]
Sorry, I didn't mean to come across as rude and that was an insensitive way of phrasing things. I understand why it was added (I read through all the related issues) but I'm just putting down my thoughts on its overall utility and how it fits with everything else in the library. Forgive me, but [you've personally turned down requests](https://github.com/nodatime/nodatime/issues/1192#issuecomment-828974524) for new combination types that don't surface any need that can't be reasonably achieved with what's already in the library.

As I said, "I don't think anything should be done about it," and I'm certainly not saying if I have no use for it then it shouldn't exist. I'm just sharing my technical feedback as a long-time user of this great library about something that I found to be extremely confusing because it led me down a wild goose chase thinking `OffsetDate` exposed some functionality it didn't actually provide (not due to any deficiency in the library but because such functionality can't reasonably exist given the limited knowledge a date + offset provides in the first place).

Again, my apologies if I came off strongly and I absolutely do not mean to disparage your work or this library. 💐
